gtdynamics/optimizer/ConstraintManifold.h

```diff
@@ -0,0 +1,162 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file  ConstraintManifold.h
+ * @brief Manifold representing variables satisfying constraints.
+ * @author: Yetong Zhang
+ */
+
+#pragma once
+
+#include <gtdynamics/optimizer/ConnectedComponent.h>
+
+namespace gtsam {
+
+/** Manifold representing constraint-connected component. Any element on the
+ * manifold is the values of variables in CCC satisfying the constraints, e.g.,
+ * {X : h(X)=0}. */
+class ConstraintManifold {
+ public:
+  /** Parameters for constraint manifold. */
+  struct Params {
+    // Method to compute tangent space basis
+    enum BasisType { KERNEL = 0, ELIMINATION = 1, SPECIFY_VARIABLES = 2 };
+    // Method to perform retraction
+    enum RetractType { UOPT = 0, PROJ = 1, PARTIAL_PROJ = 2 };
+
+    LevenbergMarquardtParams lm_params;
+    RetractType retract_type;
+    BasisType basis_type;
+
+    using shared_ptr = boost::shared_ptr<Params>;
+    Params()
+        : lm_params(),
+          retract_type(RetractType::UOPT),
+          basis_type(BasisType::KERNEL) {}
+  };
+
+ protected:
+  Params::shared_ptr params_;
+  ConnectedComponent::shared_ptr cc_;
+  size_t base_dim_;
+  size_t constraint_dim_;
+  size_t dim_;
+  gtsam::Values values_;
+  gtsam::Matrix basis_;
+  std::map<Key, size_t> var_location_;
+  std::map<Key, size_t> var_dim_;
+  KeyVector basis_keys_;
+
+ public:
+  enum { dimension = Eigen::Dynamic };
+
+  typedef OptionalJacobian<Eigen::Dynamic, Eigen::Dynamic> ChartJacobian;
+
+  /** Constructor from Connected Component.
+   * @param cc      constraint-connected component
+   * @param values  values of variable in the connected component
+   * @param retract_init  whether to perform retract in initialization
+   * @param construct_basis compute basis on initialization
+   */
+  ConstraintManifold(
+      const ConnectedComponent::shared_ptr& cc, const gtsam::Values& values,
+      const Params::shared_ptr& params = boost::make_shared<Params>(),
+      const bool retract_init = true, const bool construct_basis = true,
```

Comment:
Primitive types don't need to be const since they are a "use and throw" when constructing the object.

Reply:
Done